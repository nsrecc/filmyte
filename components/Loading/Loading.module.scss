@use "styles/mixins" as m;

.loadingWrapper {
  @include m.maxWidth;
  display: flex;
  justify-content: center;
  align-items: center;
}
