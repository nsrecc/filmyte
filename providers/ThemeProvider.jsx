import React, {
  useState, useEffect, useContext, createContext,
} from 'react';
import PropTypes from 'prop-types';
import includes from 'lodash/includes';
import styles from './ThemeProvider.module.scss';

// List of accepted themes
const THEME_LIST = ['light', 'dark'];

// Create ThemeContext with default object to be accessed in any nested component as needed.
const ThemeContext = createContext({
  theme: 'light',
  themeIsLight: true,
  toggleTheme: () => {},
});

// Export useThemeContext to retrieve theme state object via useContext hook at any nested
// components in order to update ThemeContext theme state there.
export const useThemeContext = () => useContext(ThemeContext);

/**
 * Custom hook useLocalTheme
 * - Create theme state with defaultTheme
 * - Updates theme in local storage and state
 * - Gets theme when ThemeProvider mounts
 * @param {string} defaultTheme - default theme, should be 'light'
 */
const useLocalTheme = (defaultTheme) => {
  // create theme state
  const [theme, setTheme] = useState(defaultTheme);

  // take new theme and set in it local storage, update theme state
  const setThemeLocally = (newTheme) => {
    localStorage.setItem('theme', newTheme);
    setTheme(newTheme);
  };

  // when ThemeProvider is mounted, get theme from local storage and set it to theme state
  useEffect(() => {
    const retrievedLocalTheme = localStorage.getItem('theme');

    // ensure only accepted themes can be set to theme state
    if (includes(THEME_LIST, retrievedLocalTheme)) setTheme(retrievedLocalTheme);
  }, []);

  return [theme, setThemeLocally];
};

/**
 * ThemeProvider wraps the application in _app.jsx with passed values:
 * - theme state
 * - theme check
 * - toggle theme function
 */
export const ThemeProvider = ({ children }) => {
  const [theme, setThemeLocally] = useLocalTheme('light');

  const themeIsLight = theme === 'light';

  // sets new theme using setThemeLocally from custom hook useLocalTheme
  const toggleTheme = () => {
    return themeIsLight ? setThemeLocally('dark') : setThemeLocally('light');
  };

  // div element here should contain 'app' style class as it's a direct parent to <main>
  return (
    <ThemeContext.Provider value={{ theme, themeIsLight, toggleTheme }}>
      <div className={`app ${styles.theme} ${themeIsLight ? styles.light : styles.dark}`}>
        {children}
      </div>
    </ThemeContext.Provider>
  );
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
