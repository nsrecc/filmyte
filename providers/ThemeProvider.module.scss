@use "styles/variables" as v;

/* Theme Colors */
// Use Sass Interpolation to use Sass variables with CSS Variables:
// https://sass-lang.com/documentation/breaking-changes/css-vars 
.light {
  --theme-page-bg-primary: #{v.$gray-90};
  --theme-page-bg-secondary: #{v.$gray-80};
  --theme-page-text: #{v.$black-pure};
  --theme-header-bg: #{v.$filmyte-rajah};
}

.dark {
  --theme-page-bg-primary: #{v.$gray-10};
  --theme-page-bg-secondary: #{v.$gray-20};
  --theme-page-text: #{v.$white-pure};
  --theme-header-bg: #{v.$filmyte-mulled-wine};
}

/* Theme Sass */
.theme {
  background-color: var(--theme-page-bg-primary);
  color: var(--theme-page-text);
}
