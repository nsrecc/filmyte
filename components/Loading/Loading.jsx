import React from 'react';
import { useThemeContext } from 'providers/ThemeProvider';
import PulseLoader from 'react-spinners/PulseLoader';
import styles from './Loading.module.scss';

const Loading = () => {
  const { themeIsLight } = useThemeContext();
  const loaderColor = themeIsLight ? '#000000' : '#ffffff';

  return (
    <div className={styles.loadingWrapper}>
      <PulseLoader color={loaderColor} size="1rem" />
      {/* TO DO: Accessibility sr-only element and test <span></span> */}
    </div>
  );
};

export default Loading;
