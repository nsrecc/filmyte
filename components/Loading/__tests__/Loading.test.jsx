/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { useThemeContext } from 'providers/ThemeProvider';
import Loading from '../Loading';

jest.mock('providers/ThemeProvider', () => ({
  useThemeContext: jest.fn().mockName('useThemeContext'),
}));

describe('Loading', () => {
  beforeEach(() => {
    useThemeContext.mockReturnValue({ themeIsLight: true });
  });

  // after each test, clear the calls and instances of all mocks
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<Loading />);
    expect(wrapper).toMatchSnapshot();
  });

  it('contains react-spinners Loader element', () => {
    const wrapper = shallow(<Loading />);
    expect(wrapper.find('Loader').exists()).toBe(true);
  });

  describe('Loading with light theme', () => {
    it('contains react-spinners Loader element with prop color black', () => {
      const wrapper = shallow(<Loading />);
      expect(wrapper.find('Loader').prop('color')).toBe('#000000');
    });
  });

  describe('Loading with dark theme', () => {
    beforeEach(() => {
      useThemeContext.mockReturnValue({ themeIsLight: false });
    });

    it('contains react-spinners Loader element with prop color white', () => {
      const wrapper = shallow(<Loading />);
      expect(wrapper.find('Loader').prop('color')).toBe('#ffffff');
    });
  });
});
