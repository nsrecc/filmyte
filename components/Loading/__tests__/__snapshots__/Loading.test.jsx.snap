// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading matches snapshot 1`] = `
<div
  className="loadingWrapper"
>
  <Loader
    color="#000000"
    css=""
    loading={true}
    margin={2}
    size="1rem"
  />
</div>
`;
